### PR TITLE
Embedder Asset Provider

### PR DESCRIPTION
**Pull request by ds84182**
New embedder API for allowing embedders to handle asset requests. This is working towards the option for File I/O-less Engine builds. Some embedders may have custom filesystem APIs that are incompatible with fml's File & Directory APIs. Instead of trying to put a square peg in a round hole, allow embedders to handle I/O instead.

flutter/flutter#93838

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jmagman** [comment]
`Mac Unopt` error is at https://ci.chromium.org/p/flutter/builders/try/Mac%20Unopt/6180
Rebase onto https://github.com/flutter/engine/pull/29668

**Comment by ds84182** [comment]
This PR is ready for review.

File: shell/platform/embedder/embedder_mapping.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/embedder/embedder_mapping.h"
+#include "flutter/shell/platform/embedder/embedder_struct_macros.h"
+
+namespace flutter {
+
+class EmbedderMapping final : public fml::Mapping {
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder_mapping.cc:10]: This class is unnecessary and can be replaced by `fml::NonOwnedMapping`. You may construct the `ReleaseProc` using captures. So, something like:

```
release_proc = [destruction_callback, user_data](const uint8_t* data, size_t size {
  destruction_callback(user_data);
};
```

File: shell/platform/embedder/embedder_asset_resolver.cc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/embedder/embedder_asset_resolver.h"
+#include "flutter/shell/platform/embedder/embedder_struct_macros.h"
+
+namespace flutter {
+
+class EmbedderAssetResolver final : public flutter::AssetResolver {
+ public:
+  explicit EmbedderAssetResolver(void* user_data,
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder_asset_resolver.cc:12]: The C style API must be limited to `embedder.h` and `embedder.cc`. Prefer an `std::function` here. The baton can be captured. Then, you don't have to have separate ivars for the baton and the callback. The engine doesn't care about the baton ever anyway.

File: shell/platform/embedder/embedder_asset_resolver.cc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                                 FlutterAssetResolverGetAssetCallback get_asset)
+      : user_data_(user_data), get_asset_(get_asset) {}
+
+  ~EmbedderAssetResolver() override {}
+
+  // |AssetResolver|
+  bool IsValid() const override { return true; }
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder_asset_resolver.cc:19]: Perhaps check if the callback here is non-null?

File: shell/platform/embedder/embedder_asset_resolver.cc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  // |AssetResolver|
+  AssetResolverType GetType() const override {
+    return AssetResolverType::kEmbedderAssetResolver;
+  }
+
+  // |AssetResolver|
+  std::unique_ptr<fml::Mapping> GetAsMapping(
+      const std::string& asset_name) const override {
+    if (!get_asset_) {
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder_asset_resolver.cc:32]: This isn't how the the other asset resolvers behave. If the asset cannot be found, `nullptr` must be returned which allows the lookup to move on to the next resolver. I don't think we have a mechanism built-in to the engine to short circuit this mechanism (nor should we have one).

File: shell/platform/embedder/embedder_asset_resolver.cc (hunk 2 above)
Comment (chinmaygarde) [reply shell/platform/embedder/embedder_asset_resolver.cc:12]: Better yet, make the caller return a valid mapping.

So something like: `using Callback = std::function<std::unique_ptr<fml::Mapping>(const std::string& asset_name)>;`

File: shell/platform/embedder/embedder_asset_resolver.cc [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace flutter {
+
+class EmbedderAssetResolver final : public flutter::AssetResolver {
+ public:
+  explicit EmbedderAssetResolver(void* user_data,
+                                 FlutterAssetResolverGetAssetCallback get_asset)
+      : user_data_(user_data), get_asset_(get_asset) {}
+
+  ~EmbedderAssetResolver() override {}
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder_asset_resolver.cc:16]: Nit `= default;` and specify the interface for the override.

**Review by chinmaygarde — CHANGES_REQUESTED** [review]
The public API is prone to misuse in the following cases:

* If the embedder creates and caches mapping objects (something that is very reasonable for embedders to do), the second instance of the engine asking the embedder to fetch the asset will lead to the mapping being held by two unique pointers (in the asset resolver). A release by the first instance will lead to an double free.
* If the mapping is created by the embedder but the embedder never returns it via a callback, they have no way of collecting the handle and it will be leaked.

In a C API, it's usually suspicious when a `Create` call is not balanced by a subsequent `Collect`. I suggest adding the two with a handle to a struct that has shared ownership of the mapping being returned to the embedder. Then, in the resolver, the engine can just ref the handle.

Other construction issues and nits added inline. Thanks and sorry about the delayed review.

**Comment by ds84182** [comment]
@chinmaygarde 

> If the embedder creates and caches mapping objects (something that is very reasonable for embedders to do), the second instance of the engine asking the embedder to fetch the asset will lead to the mapping being held by two unique pointers (in the asset resolver). A release by the first instance will lead to an double free.

The documentation on the API specifies that the embedder should always return fresh mapping objects from the callback. Otherwise yes they will encounter a double-free. If the embedder wants to share the underlying data with multiple mappings, they should maintain some sort of reference count. Should the documentation outline this more explicitly?

As a quick example of usage (from Rust for brevity, also possible with std::shared_ptr):

```rust
let my_shared_data: Arc<[u8; N]> = Arc::new(b"my file bytes here");

// Creating mappings for the shared data:
let mut create_info = FlutterEngineMappingCreateInfo::new();

create_info.data = my_shared_data.as_ptr();
create_info.data_size = my_shared_data.len();

// Increment the reference count and turn into a raw ptr
create_info.user_data = Arc::into_raw(Arc::clone(my_shared_data));
unsafe extern "C" fn destroy(user_data: *mut c_void) {
  // Turn raw ptr into the associated Arc and drop to decrement reference count
  Arc::from_raw(user_data as *const [u8; N]);
}
create_info.destruction_callback = destroy;

let mapping = FlutterEngineCreateMapping(&create_info);
```

> If the mapping is created by the embedder but the embedder never returns it via a callback, they have no way of collecting the handle and it will be leaked.

Good call, I'll add a destroy method for the mappings. It also makes sense if we start handing mappings from the engine -> embedder. Would it be worthwhile to also add data + size getters to fully flesh out the API?

File: shell/platform/embedder/embedder_asset_resolver.cc (hunk 4 above)
Comment (ds84182) [reply shell/platform/embedder/embedder_asset_resolver.cc:31]: Gotcha, while the previous one was equivalent to returning nullptr, I changed it to `return nullptr` so it's more clear.

**Comment by ds84182** [comment]
I've updated the `FlutterMapping` API to be a bit more general. The comments on the mapping-related calls explain the required ownership semantics.

I was debating whether it's worthwhile to change the interface some more to be more explicit about the ownership requirements. For example using a different opaque struct for Engine-owned mappings vs Embedder-owned mappings, and having a call to "give ownership" of an Embedder mapping to the Engine. Something like:

```c
typedef struct FlutterEngineOwnedMapping* FlutterEngineOwnedMapping;
typedef struct FlutterEmbedderOwnedMapping* FlutterEmbedderOwnedMapping;

// Releases an embedder-owned mapping to the engine. This is used to give the mapping to the engine inside of
// AssetResolver->get_asset.
// Internally just a pointer cast, but may do some additional debug-only book-keeping in the future.
FlutterEngineOwnedMapping FlutterReleaseMappingOwnership(FlutterEmbedderOwnedMapping mapping);

void FlutterCreateMapping(..., FlutterEmbedderOwnedMapping* out);
void FlutterDestroyMapping(FlutterEmbedderOwnedMapping mapping);
const uint8_t* FlutterGetMappingData(FlutterEmbedderOwnedMapping mapping, ...);
```

---

If we want to also have an engine-provided shared mapping, I think it should have a different interface. e.g.

```c
typedef struct FlutterSharedMapping* FlutterSharedMapping;
void FlutterCreateSharedMapping(const FlutterSharedMappingCreateInfo* create_info, FlutterSharedMapping* out);

// Increases refcount, can be directly returned from AssetResolver->get_asset
FlutterEngineOwnedMapping FlutterMakeEngineMappingFromShared(FlutterSharedMapping mapping);

// The mapping may stay alive if there are any mappings owned by the engine.
// New mappings cannot be created from the shared mapping after this point.
void FlutterDestroySharedMapping(FlutterSharedMapping mapping);
```

However, I think this is tangential to this PR and can be implemented by embedders with the current API.

**Comment by zanderso** [comment]
What is the status of this PR?

**Comment by ds84182** [comment]
@zanderso Currently on the backburner. Focus has changed on my side somewhat, but I can spend a bit of time solving the merge conflict.

Other than that I think there's still some open discussion from @chinmaygarde about how FlutterMapping is exposed to the Embedder.

**Comment by zanderso** [comment]
Marking this WIP, but cc @iskakaushik regarding an issue from @akbiggs in https://github.com/flutter/flutter/issues/100640.

**Comment by Hixie** [comment]
I'm going to close this one for now per @ds84182; they have the branch locally and will submit a new PR when ready.

**Pull request closed without merging**